test/distributed/checkpoint/test_fsdp_model_state.py

```diff
@@ -0,0 +1,102 @@
+import torch.distributed as dist
+
+from torch.distributed.checkpoint.default_planner import (
+    DefaultSavePlanner,
+    DefaultLoadPlanner,
+)
+
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    with_comms,
+)
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_utils import run_tests
+from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
+
+
+class FsdpModelStateCheckpoint(DTensorTestBase):
+    def _test_fsdp_model_state(self, process_group) -> None:
+        CHECKPOINT_DIR = self.temp_dir
+
+        model = FSDP(torch.nn.Linear(8, 8, device="meta"))
+        model(torch.rand(8, 8, device=dist.get_rank())).sum().backward()
+
+        with FSDP.state_dict_type(model, StateDictType.SHARDED_STATE_DICT):
+            state_dict = {
+                "model": model.state_dict(),
+            }
+
+            dist_cp.save_state_dict(
+                state_dict=state_dict,
+                storage_writer=dist_cp.FileSystemWriter(CHECKPOINT_DIR),
+                planner=DefaultSavePlanner(
+                    flatten_state_dict=True,
+                    flatten_sharded_tensors=True,
+                ),
+            )
+
+        model_2 = FSDP(
+            torch.nn.Linear(8, 8, device="meta"), process_group=process_group
+        )
+
+        with FSDP.summon_full_params(model):
+            with FSDP.summon_full_params(model_2):
+                self.assertNotEqual(model.weight, model_2.weight)
+                self.assertNotEqual(model.bias, model_2.bias)
+
+        # now load the model and ensure the values are the same
+        with FSDP.state_dict_type(model_2, StateDictType.SHARDED_STATE_DICT):
+            state_dict = {
+                "model": model_2.state_dict(),
+            }
+
+            dist_cp.load_state_dict(
+                state_dict=state_dict,
+                storage_reader=dist_cp.FileSystemReader(CHECKPOINT_DIR),
+                planner=DefaultLoadPlanner(
+                    flatten_state_dict=True,
+                    flatten_sharded_tensors=True,
+                ),
+            )
+            model_2.load_state_dict(state_dict["model"])
+
+        with FSDP.summon_full_params(model):
+            with FSDP.summon_full_params(model_2):
+                self.assertEqual(model.weight, model_2.weight)
+                self.assertEqual(model.bias, model_2.bias)
+
+    @with_comms
+    @skip_if_lt_x_gpu(4)
+    @with_temp_dir
+    def test_fsdp_model_state_no_resharding(self):
+        self._test_fsdp_model_state(process_group=None)
+
+    def _create_new_dist_group(self):
+        # create new fsdp group for resharding
+        fsdp_0 = dist.new_group(ranks=[0, 2])
+        fsdp_1 = dist.new_group(ranks=[1, 3])
+        if dist.get_rank() % 2 == 0:
+            my_fsdp = fsdp_0
+        else:
+            my_fsdp = fsdp_1
+
+        return my_fsdp
```

Comment:
we skip the test when # of GPU is less than 4. What if we use 8 GPUs for test. Shall we not hard code this logic?

Reply:
Thanks! Nice catch! Changed it to form two new dist group (1 for odd number rank and 1 for even number rank).